package com.gildedtros;

import java.util.Arrays;

class GildedTros {
    Item[] items;

    private final String[] legendaryNameList = {"B-DAWG Keychain"};

    public GildedTros(Item[] items) {
        this.items = items;
    }

    public void updateQuality() {
        for (int i = 0; i < items.length; i++) {
            if (!items[i].name.equals("Good Wine")
                    && !items[i].name.equals("Backstage passes for Re:Factor")
                    && !items[i].name.equals("Backstage passes for HAXX"))
            {
                if (items[i].quality > 0) {
                    if (!items[i].name.equals("B-DAWG Keychain")) {
                        items[i].quality = items[i].quality - 1;
                    }
                }
            } else {
                if (items[i].quality < 50) {
                    items[i].quality = items[i].quality + 1;

                    if (items[i].name.equals("Backstage passes for Re:Factor") || items[i].name.equals("Backstage passes for HAXX") ) {
                        if (items[i].sellIn < 11) {
                            if (items[i].quality < 50) {
                                items[i].quality = items[i].quality + 1;
                            }
                        }

                        if (items[i].sellIn < 6) {
                            if (items[i].quality < 50) {
                                items[i].quality = items[i].quality + 1;
                            }
                        }
                    }
                }
            }

            if (!items[i].name.equals("B-DAWG Keychain")) {
                items[i].sellIn = items[i].sellIn - 1;
            }

            if (items[i].sellIn < 0) {
                if (!items[i].name.equals("Good Wine")) {
                    if (!items[i].name.equals("Backstage passes for Re:Factor") && !items[i].name.equals("Backstage passes for HAXX")) {
                        if (items[i].quality > 0) {
                            if (!items[i].name.equals("B-DAWG Keychain")) {
                                items[i].quality = items[i].quality - 1;
                            }
                        }
                    } else {
                        items[i].quality = items[i].quality - items[i].quality;
                    }
                } else {
                    if (items[i].quality < 50) {
                        items[i].quality = items[i].quality + 1;
                    }
                }
            }
        }
    }

    public void updateQualityV2() {
        for (int i = 0; i < items.length; i++) {
            if(Arrays.asList(legendaryNameList).contains(items[i].name)) {
                items[i].quality = 80;
                items[i].sellIn = 0;
                continue;
            }
        }
    }

    private void increaseQuality(int itemIdx, int amount) {
        items[itemIdx].quality += amount;
        if (items[itemIdx].quality > 50) {
            items[itemIdx].quality = 50;
        }
    }

    private void decreaseQuality(int itemIdx, int amount) {
        items[itemIdx].quality -= amount;
        if (items[itemIdx].quality < 0) {
            items[itemIdx].quality = 0;
        }
    }
}